[stdlib] Speed up bridged Dictionary instances

lorentey opened the pull request:
`Dictionary`’s native storage classes and `_SwiftDeferredNSDictionary` override `-[NSDictionary getObjects:andKeys:]` instead of its safer replacement, `-[NSDictionary getObjects:andKeys:count:]`.

Overriding the correct method will considerably speed up some Cocoa operations on bridged dictionaries.

rdar://problem/39285882

lorentey committed:
[stdlib] Dictionary: Override getObjects:andKeys:count:

Dictionary’s native storage classes and _SwiftDeferredNSDictionary override -[NSDictionary getObjects:andKeys:] instead of its safer replacement, -[NSDictionary getObjects:andKeys:count:].

Overriding the correct method will considerably speed up some Cocoa operations on bridged dictionaries.

rdar://problem/39285882

lorentey commented:
@swift-ci please test

lorentey commented:
I don't think we have benchmarks for bridged dictionaries yet, but just in case:

@swift-ci please benchmark

lorentey committed:
[benchmark] Add NSDictionary benchmarks for dictionaries bridged from Swift

jtbandes commented on stdlib/public/core/Dictionary.swift:
```diff
@@ -2130,28 +2133,33 @@ final internal class _HashableTypedNativeDictionaryStorage<Key: Hashable, Value>
         for (offset: i, element: (key: key, value: val)) in full.enumerated() {
           unmanagedObjects[i] = _bridgeAnythingToObjectiveC(val)
           unmanagedKeys[i] = _bridgeAnythingToObjectiveC(key)
+          guard i < count else { break }
```
Sorry for the drive-by comment, but shouldn’t this be checked _before_ assigning at position `i`? (Same applies below)

lorentey commented:
I added new microbenchmarks to help us track performance. Copying elements in bulk out of an `_SwiftDeferredNSDictionary` is now ~5 times faster:

<details open>
  <summary>Improvement (1)</summary>
  
TEST                              | OLD | NEW | DELTA  | SPEEDUP   
---                               | --- | --- | ---    | ---       
DictionaryBridgeToObjC_BulkAccess | 699 | 143 | -79.5% | **4.89x** 

</details>

<details open>
  <summary>No Changes (2)</summary>
  
TEST                              | OLD | NEW | DELTA  | SPEEDUP   
---                               | --- | --- | ---    | ---       
DictionaryBridgeToObjC_Access     | 774 | 760 | -1.8%  | 1.02x (?) 
DictionaryBridgeToObjC_Bridge     | 18  | 18  | +0.0%  | 1.00x     

</details>

lorentey committed:
[test] check getObjects:andKeys:count: implementations in bridged Dictionaries

lorentey committed:
[stdlib] Dictionary: Fix getObjects:andKeys:count: implementations

lorentey committed:
[stdlib] _NSDictionary: Replace getObjects:andKeys: with getObjects:andKeys:count:

lorentey committed:
[stdlib] StdlibUnittestFoundationExtras: fix warnings

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci test

michael-lehew commented on stdlib/public/core/Dictionary.swift:
```diff
@@ -2961,6 +2975,7 @@ final internal class _SwiftDeferredNSDictionary<Key: Hashable, Value>
         // keys null, objects nonnull
         for position in 0..<bucketCount {
           if bridgedBuffer.isInitializedEntry(at: position) {
+            guard i < count else { break }
```
you could eke out a one less loop iteration by checking this after you increment i -- and adding an early out above for count == 0

lorentey replied:
Good point! It can often be more than one iteration, too, since the loop skips over empty buckets.

moiseev commented:
:shipit:

lorentey committed:
[stdlib] Dictionary: Optimize loops in getObjects:andKeys:count: implementations

This breaks out of the loop immediately when the last slot has been filled in the output buffer, skipping a final sequence of iterations over empty buckets.

lancep reviewed (approved):
LGTM! 🍔

lorentey commented:
@swift-ci test and merge
